### PR TITLE
[script.audiobooks] 1.4.0

### PR DESCRIPTION
**Pull request by robwebset**
Changelog:

v1.4.0 (2016-03-19)
- Order Books by title if available, otherwise file name
- Support folder.jpg and cover.jpg files
- Improve end of book detection for multiple file books
- Prevent the lyrics being searched for audiobooks

For Releases Gotham and beyond

### CODE REVIEW

File: script.audiobooks/plugin.py [hunk 1]
```diff
@@ -1,418 +1,428 @@
...
+    # Creates a URL for a directory
+    def _build_url(self, query):
+        return self.base_url + '?' + urllib.urlencode(query)
+
+    # Show all the EBooks that are in the eBook directory
+    def showAudiobooks(self, directory=None):
+        # Get the setting for the audio book directory
+        audioBookFolder = Settings.getAudioBookFolder()
+
+        if audioBookFolder in [None, ""]:
```
Comment (phil65) [inline script.audiobooks/plugin.py]: you can probably use "if not audioBookFolder:" for these cases.

File: script.audiobooks/plugin.py [hunk 2]
```diff
@@ -1,418 +1,428 @@
...
+        if directory not in [None, ""]:
+            audioBookFolder = directory
+
+        dirs, files = xbmcvfs.listdir(audioBookFolder)
+        files.sort()
+        dirs.sort()
+
+        bookDirs = []
+        # For each directory list allow the user to navigate into it
+        for dir in dirs:
```
Comment (phil65) [inline script.audiobooks/plugin.py]: dir is a builtin function, it´s recommended to not use those for variable names.

File: script.audiobooks/plugin.py [hunk 3]
```diff
@@ -1,418 +1,428 @@
...
+    def _isAudioBookDir(self, fullDir):
+        # Check to see if this directory contains audio files (non m4b), if it does then we construct the
+        # book using each audio file file as a chapter
+        dirs, files = xbmcvfs.listdir(fullDir)
+
+        containsMP3 = False
+        for aFile in files:
+            if Settings.isPlainAudioFile(aFile):
+                log("AudioBooksPlugin: Directory contains MP3 files: %s" % fullDir)
+                containsMP3 = True
```
Comment (phil65) [inline script.audiobooks/plugin.py]: you could also put a "return True" here directly.

File: script.audiobooks/plugin.py [hunk 4]
```diff
@@ -1,418 +1,428 @@
...
+                    durationEntry = 1
+                # Use the start time for the duration display as that will show
+                # how far through the book the chapter is
+                li.setInfo('music', {'Duration': durationEntry})
+
+            li.setProperty("Fanart_Image", __fanart__)
+            li.addContextMenuItems([], replaceItems=True)
+            xbmcplugin.addDirectoryItem(handle=self.addon_handle, url=url, listitem=li, isFolder=False)
+
+        del audioBookHandler
```
Comment (phil65) [inline script.audiobooks/plugin.py]: "del" doesnt really do anything here :)  python handles garbage collection itself.

File: script.audiobooks/plugin.py [hunk 5]
```diff
@@ -1,418 +1,428 @@
...
+
+    def play(self, fullpath, startTime=0, chapter=0):
+        log("AudioBooksPlugin: Playing %s" % fullpath)
+
+        audioBookHandler = AudioBookHandler.createHandler(fullpath)
+
+        bookPlayer = BookPlayer()
+        bookPlayer.playAudioBook(audioBookHandler, startTime, chapter)
+        del bookPlayer
+        del audioBookHandler
```
Comment (phil65) [inline script.audiobooks/plugin.py]: same for these 2

File: script.audiobooks/plugin.py [hunk 6]
```diff
@@ -1,418 +1,428 @@
...
+        cmd = self._build_url({'mode': 'clear', 'filename': bookHandle.getFile(True)})
+        ctxtMenu.append((__addon__.getLocalizedString(32011), 'RunPlugin(%s)' % cmd))
+
+        return ctxtMenu
+
+    def progress(self, fullpath, isComplete=True, startTime=0):
+        # At the moment the only time progress is called is to mark as complete
+        audiobookDB = AudioBooksDB()
+        audiobookDB.setPosition(fullpath, startTime, isComplete)
+        del audiobookDB
```
Comment (phil65) [inline script.audiobooks/plugin.py]: ..and here

File: script.audiobooks/plugin.py [hunk 7]
```diff
@@ -1,418 +1,428 @@
...
+
+        xbmc.executebuiltin("Container.Refresh")
+
+    def clear(self, fullpath):
+        log("AudioBooksPlugin: Clearing history for %s" % fullpath)
+        # Remove the item from the database, it will then be rescanned
+        audiobookDB = AudioBooksDB()
+        audiobookDB.deleteAudioBook(fullpath)
+        del audiobookDB
```
Comment (phil65) [inline script.audiobooks/plugin.py]: ..and here

File: script.audiobooks/plugin.py [hunk 8]
```diff
@@ -1,418 +1,428 @@
...
+            del menuNav
+
+    elif mode[0] == 'chapters':
+        log("AudioBooksPlugin: Mode is CHAPTERS")
+
+        # Get the actual folder that was navigated to
+        filename = args.get('filename', None)
+        cover = args.get('cover', None)
+
+        if (cover is not None) and (len(cover) > 0):
```
Comment (phil65) [inline script.audiobooks/plugin.py]: "if cover:" should be enough

File: script.audiobooks/resources/lib/audiobook.py [hunk 9]
```diff
@@ -1,661 +1,690 @@
...
+        log("AudioBookHandler: Loading audio book %s (%s)" % (self.filePath, self.fileName))
+
+        # Check to see if we already have an image available
+        self.coverImage = self._getExistingCoverImage()
+
+        # Check in the database to see if this audio book is already recorded
+        audiobookDB = AudioBooksDB()
+        audiobookDetails = audiobookDB.getAudioBookDetails(self.filePath)
+
+        if audiobookDetails not in [None, ""]:
```
Comment (phil65) [inline script.audiobooks/resources/lib/audiobook.py]: see above

**Comment by phil65** [comment]
Picked some minors, feel free to fix them or not.
unicode / str encoding/decoding also looks quite wonky. 
Apart from that, all fine. Thx.
(just quick code review, no runtime test)

**Comment by robwebset** [comment]
Thanks @phil65 for taking a look

I have updated the user of "dir" so that it doesn't clash - I think several of the others are coding style things.

Yes the the unicode is horrid - I hate Python character encoding!

Thanks, Rob

**Comment by phil65** [comment]
thx!

**Pull request merged**